packages/shared/lib/services/connection.service.ts

```diff
@@ -846,150 +847,151 @@ class ConnectionService {
-                    const { success, response: connection } = await this.getConnection(connectionId, providerConfig.unique_key, environmentId);
-                    if (success && connection) {
-                        const shouldRefresh = await this.shouldRefreshCredentials(
-                            connection,
-                            connection.credentials as OAuth2Credentials,
-                            providerConfig,
-                            template,
-                            false
-                        );
-                        if (!shouldRefresh) {
-                            return {
-                                success: true,
-                                error: null,
-                                response: {
-                                    refreshed: true,
-                                    credentials: connection.credentials as OAuth2Credentials | AppCredentials | AppStoreCredentials | OAuth2ClientCredentials
-                                }
-                            };
-                        }
-                    }
-                    throw err;
-                }
+            if (!success || !connection) {
+                throw error;
+            }
 
-                const { success, error, response: newCredentials } = await this.getNewCredentials(freshConnection, providerConfig, template);
-                if (!success || !newCredentials) {
-                    await telemetry.log(LogTypes.AUTH_TOKEN_REFRESH_FAILURE, `Token refresh failed, ${error?.message}`, LogActionEnum.AUTH, {
-                        environmentId: String(environment_id),
-                        connectionId,
-                        providerConfigKey,
-                        provider: providerConfig.provider,
-                        level: 'error'
-                    });
+            const shouldRefresh = await this.shouldRefreshCredentials(
+                connection,
+                connection.credentials as OAuth2Credentials,
+                providerConfig,
+                template,
+                instantRefresh
+            );
 
-                    return { success, error, response: null };
-                }
+            return { connection, shouldRefresh };
+        };
 
-                freshConnection.credentials = newCredentials;
-                await this.updateConnection(freshConnection);
+        // Check if the credentials need to be refreshed
+        // If not, return the current credentials as they are still valid
+        const { connection, shouldRefresh } = await getConnection();
+        if (!shouldRefresh) {
+            return {
+                success: true,
+                error: null,
+                response: {
+                    refreshed: false,
+                    credentials: connection.credentials as OAuth2Credentials | AppCredentials | AppStoreCredentials | OAuth2ClientCredentials
+                }
+            };
+        }
```

Comment:
this early call/return is incorrect, if a refresh is in progress the credentials can be invalid just after the return

Reply:
good point. that's another problem that I haven't address, even though in practice I am not sure it can happen.